ironfish/src/indexer/minedBlocksIndexer.ts

```diff
@@ -0,0 +1,268 @@
+    chain: Blockchain
+    logger?: Logger
+  }) {
+    this.files = files
+    this.location = location
+    this.database = createDB({ location })
+    this.accounts = accounts
+    this.logger = logger
+    this.chain = chain
+    this.isOpen = false
+    this.isStarted = false
+
+    this.meta = this.database.addStore<{
+      key: keyof MinedBlocksDBMeta
+      value: MinedBlocksDBMeta[keyof MinedBlocksDBMeta]
+    }>({
+      name: 'minedBlocksDBMeta',
+      keyEncoding: new StringEncoding<keyof MinedBlocksDBMeta>(),
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.minedBlocks = this.database.addStore<{ key: string; value: MinedBlock }>({
+      name: 'minedBlocks',
+      keyEncoding: new StringEncoding(),
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.accountToBlockHashes = this.database.addStore<{ key: string; value: string[] }>({
+      name: 'accountToBlockHashes',
+      keyEncoding: new StringEncoding(),
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.chainProcessor = new ChainProcessor({ logger, chain, head: null })
+
+    this.chainProcessor.onAdd.on(async (header) => {
+      const block = await this.chain.getBlock(header)
+      Assert.isNotNull(block)
+
+      const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+      await this.database.transaction(async (tx) => {
+        if (account) {
+          await this.minedBlocks.put(
+            header.hash.toString('hex'),
+            {
+              main: true,
+              hash: header.hash.toString('hex'),
+              sequence: header.sequence,
+              account: account.name,
+              minersFee: Number(header.minersFee),
+            },
+            tx,
+          )
+          const hashes: string[] = (await this.accountToBlockHashes.get(account.name, tx)) || []
+          hashes.concat(block.header.hash.toString('hex'))
+          await this.accountToBlockHashes.put(account.name, hashes, tx)
+        }
+      })
+
+      await this.updateHeadHash(header.hash)
+    })
+
+    this.chainProcessor.onRemove.on(async (header) => {
+      if (await this.minedBlocks.has(header.hash.toString('hex'))) {
+        const block = await this.chain.getBlock(header)
+        Assert.isNotNull(block)
+
+        const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+        if (account) {
+          const minedBlock = await this.minedBlocks.get(header.hash.toString('hex'))
+          if (minedBlock) {
+            minedBlock.main = false
+            await this.minedBlocks.put(header.hash.toString('hex'), minedBlock)
+          }
+        }
+      }
+
+      await this.updateHeadHash(header.previousBlockHash)
+    })
+
+    this.accounts.onAccountCreated.on((_account, imported) => {
+      if (imported) {
+        this.chainProcessor.hash = null
+      }
+    })
+
+    this.accounts.onAccountRemoved.on(async (accountName) => {
+      await this.database.transaction(async (tx) => {
```

Comment:
Is this just used as a lock to avoid doing this while the chain processor is running?

Reply:
Essentially, yes. I'm ensuring that it's an atomic change.